File: web/src/crypto/identity.ts
import { decrypt, encrypt, exportKey, generateKey, importKey, type Sealed } from './aead';
import { base32, concat, group, utf8 } from './bytes';

/**
 * A member holds two P-256 keypairs: one for key agreement, one for signatures.
 * WebCrypto binds a key to a single algorithm, and reusing one EC key across ECDH and
 * ECDSA is exactly the kind of shortcut that turns into a break later.
 *
 * Both public keys travel in the single public_key column as one versioned blob, and
 * both private keys travel in wrapped_private_key, encrypted with the master key.
 */
export const IDENTITY_FORMAT = 0x01;

export const CURVE = 'P-256';
/** Uncompressed P-256 point: 0x04 || x || y. */
export const PUBLIC_KEY_LENGTH = 65;

const MASTER_KEY_AAD = utf8('shelf/master-key/v1');
const IDENTITY_AAD = utf8('shelf/identity/v1');

const FINGERPRINT_CHARS = 16;
const FINGERPRINT_GROUP = 4;

export interface Identity {
  /** Agreement keys: seal to `sealPublic`, open with `sealPrivate`. */
  sealPublic: CryptoKey;
  sealPrivate: CryptoKey;
  signPrivate: CryptoKey;
  signPublic: CryptoKey;
  /** The exact bytes stored in users.public_key, needed to re-derive the fingerprint. */
  publicBlob: Uint8Array;
  fingerprint: string;
}

export interface NewIdentity {
  identity: Identity;
  publicBlob: Uint8Array;
  wrappedPrivateKey: Uint8Array;
  privateKeyNonce: Uint8Array;
}

export async function generateMasterKey(): Promise<CryptoKey> {
  return generateKey();
}

/** Wraps the master key with the key derived from a passphrase or a recovery code. */
export async function wrapMasterKey(masterKey: CryptoKey, wrappingKey: CryptoKey): Promise<Sealed> {
  return encrypt(wrappingKey, await exportKey(masterKey), MASTER_KEY_AAD);
}

export async function unwrapMasterKey(sealed: Sealed, wrappingKey: CryptoKey): Promise<CryptoKey> {
  return importKey(await decrypt(wrappingKey, sealed, MASTER_KEY_AAD));
}

export async function generateIdentity(masterKey: CryptoKey): Promise<NewIdentity> {
  const seal = await crypto.subtle.generateKey({ name: 'ECDH', namedCurve: CURVE }, true, [
    'deriveBits',
  ]);
  const sign = await crypto.subtle.generateKey({ name: 'ECDSA', namedCurve: CURVE }, true, [
    'sign',
    'verify',
  ]);

  const sealPublicRaw = await rawPublic(seal.publicKey);
  const signPublicRaw = await rawPublic(sign.publicKey);
  const publicBlob = concat(Uint8Array.of(IDENTITY_FORMAT), sealPublicRaw, signPublicRaw);

  const bundle = concat(
    Uint8Array.of(IDENTITY_FORMAT),
    lengthPrefixed(new Uint8Array(await crypto.subtle.exportKey('pkcs8', seal.privateKey))),
    lengthPrefixed(new Uint8Array(await crypto.subtle.exportKey('pkcs8', sign.privateKey))),
  );

  const wrapped = await encrypt(masterKey, bundle, IDENTITY_AAD);

  return {
    identity: {
      sealPublic: seal.publicKey,
      sealPrivate: seal.privateKey,
      signPrivate: sign.privateKey,
      signPublic: sign.publicKey,
      publicBlob,
      fingerprint: await fingerprint(publicBlob),
    },
    publicBlob,
    wrappedPrivateKey: wrapped.ciphertext,
    privateKeyNonce: wrapped.nonce,
  };
}

export async function unwrapIdentity(
  publicBlob: Uint8Array,
  sealed: Sealed,
  masterKey: CryptoKey,
): Promise<Identity> {
  const { seal: sealPublicRaw, sign: signPublicRaw } = splitPublicBlob(publicBlob);
  const bundle = await decrypt(masterKey, sealed, IDENTITY_AAD);

  if (bundle[0] !== IDENTITY_FORMAT) throw new Error('unknown identity format');

  const sealPkcs8 = readLengthPrefixed(bundle, 1);
  const signPkcs8 = readLengthPrefixed(bundle, sealPkcs8.next);

  return {
    sealPublic: await importSealPublic(sealPublicRaw),
    sealPrivate: await crypto.subtle.importKey(
      'pkcs8',
      sealPkcs8.value as BufferSource,
      { name: 'ECDH', namedCurve: CURVE },
      true,
      ['deriveBits'],
    ),
    signPublic: await crypto.subtle.importKey(
      'raw',
      signPublicRaw as BufferSource,
      { name: 'ECDSA', namedCurve: CURVE },
      true,
      ['verify'],
    ),
    signPrivate: await crypto.subtle.importKey(
      'pkcs8',
      signPkcs8.value as BufferSource,
      { name: 'ECDSA', namedCurve: CURVE },
      true,
      ['sign'],
    ),
    publicBlob,
    fingerprint: await fingerprint(publicBlob),
  };
}

/** importKey rejects points off the curve, which closes the invalid-curve attack on P-256. */
export async function importSealPublic(raw: Uint8Array): Promise<CryptoKey> {
  return crypto.subtle.importKey('raw', raw as BufferSource, { name: 'ECDH', namedCurve: CURVE }, true, []);
}

export function splitPublicBlob(blob: Uint8Array): { seal: Uint8Array; sign: Uint8Array } {
  if (blob[0] !== IDENTITY_FORMAT) throw new Error('unknown public key format');

  if (blob.length !== 1 + PUBLIC_KEY_LENGTH * 2) {
    throw new Error('public key blob has an unexpected length');
  }

  return {
    seal: blob.subarray(1, 1 + PUBLIC_KEY_LENGTH),
    sign: blob.subarray(1 + PUBLIC_KEY_LENGTH),
  };
}

export async function sign(identity: Identity, payload: Uint8Array): Promise<Uint8Array> {
  const signature = await crypto.subtle.sign(
    { name: 'ECDSA', hash: 'SHA-256' },
    identity.signPrivate,
    payload as BufferSource,
  );

  return new Uint8Array(signature);
}

export async function verify(
  authorPublicBlob: Uint8Array,
  signature: Uint8Array,
  payload: Uint8Array,
): Promise<boolean> {
  const key = await crypto.subtle.importKey(
    'raw',
    splitPublicBlob(authorPublicBlob).sign as BufferSource,
    { name: 'ECDSA', namedCurve: CURVE },
    true,
    ['verify'],
  );

  return crypto.subtle.verify(
    { name: 'ECDSA', hash: 'SHA-256' },
    key,
    signature as BufferSource,
    payload as BufferSource,
  );
}

/**
 * Short, readable digest of the public blob. The server hands out public keys, so it can
 * hand out its own; comparing fingerprints out of band is what closes that gap.
 */
export async function fingerprint(publicBlob: Uint8Array): Promise<string> {
  const digest = new Uint8Array(await crypto.subtle.digest('SHA-256', publicBlob as BufferSource));

  return group(base32(digest, FINGERPRINT_CHARS), FINGERPRINT_GROUP, ' ');
}

async function rawPublic(key: CryptoKey): Promise<Uint8Array> {
  return new Uint8Array(await crypto.subtle.exportKey('raw', key));
}

function lengthPrefixed(value: Uint8Array): Uint8Array {
  const header = new Uint8Array(2);
  new DataView(header.buffer).setUint16(0, value.length, false);

  return concat(header, value);
}

function readLengthPrefixed(bundle: Uint8Array, offset: number): { value: Uint8Array; next: number } {
  if (offset + 2 > bundle.length) throw new Error('identity bundle is truncated');

  const view = new DataView(bundle.buffer, bundle.byteOffset, bundle.byteLength);
  const length = view.getUint16(offset, false);
  const start = offset + 2;

  if (start + length > bundle.length) throw new Error('identity bundle is truncated');

  return { value: bundle.subarray(start, start + length), next: start + length };
}

File: internal/envelope/identity.go
package envelope

import (
	"crypto/ecdh"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"crypto/x509"
	"encoding/binary"
	"fmt"
	"math/big"
)

const (
	// IdentityFormat is the first byte of both the public blob and the private bundle.
	IdentityFormat = 0x01

	// PublicKeyLength is an uncompressed P-256 point: 0x04 || x || y.
	PublicKeyLength = 65

	// SignatureLength is ECDSA r || s, 32 bytes each. WebCrypto produces and expects this
	// raw form, not the ASN.1 one crypto/ecdsa signs with by default.
	SignatureLength = 64
)

// SplitPublicBlob separates the two halves of the identity blob stored in users.public_key
// and in mcp_connectors.public_key. Both keys are P-256, and they are kept apart because
// reusing one EC key across ECDH and ECDSA is the kind of shortcut that breaks later.
func SplitPublicBlob(blob []byte) (seal, sign []byte, err error) {
	if len(blob) != 1+PublicKeyLength*2 {
		return nil, nil, fmt.Errorf("public key blob must be %d bytes, got %d", 1+PublicKeyLength*2, len(blob))
	}

	if blob[0] != IdentityFormat {
		return nil, nil, fmt.Errorf("unknown public key format %#x", blob[0])
	}

	return blob[1 : 1+PublicKeyLength], blob[1+PublicKeyLength:], nil
}

// ParsePrivateBundle reads the pair of PKCS#8 keys the browser packs into one blob:
// IdentityFormat, then each key behind a big-endian uint16 length, agreement key first.
func ParsePrivateBundle(bundle []byte) (*ecdh.PrivateKey, *ecdsa.PrivateKey, error) {
	if len(bundle) == 0 || bundle[0] != IdentityFormat {
		return nil, nil, fmt.Errorf("unknown identity bundle format")
	}

	sealPKCS8, next, err := readLengthPrefixed(bundle, 1)
	if err != nil {
		return nil, nil, err
	}

	signPKCS8, _, err := readLengthPrefixed(bundle, next)
	if err != nil {
		return nil, nil, err
	}

	// Both halves are id-ecPublicKey on the wire, so the agreement key parses as ECDSA and
	// is converted rather than being read as ECDH directly.
	sealKey, err := parseEC(sealPKCS8)
	if err != nil {
		return nil, nil, fmt.Errorf("parse agreement key: %w", err)
	}

	seal, err := sealKey.ECDH()
	if err != nil {
		return nil, nil, fmt.Errorf("convert agreement key: %w", err)
	}

	sign, err := parseEC(signPKCS8)
	if err != nil {
		return nil, nil, fmt.Errorf("parse signing key: %w", err)
	}

	return seal, sign, nil
}

// Verify checks a signature made by the browser, which emits r || s rather than ASN.1.
func Verify(authorPublicBlob, signature, payload []byte) bool {
	if len(signature) != SignatureLength {
		return false
	}

	_, signRaw, err := SplitPublicBlob(authorPublicBlob)
	if err != nil {
		return false
	}

	public, err := signingKey(signRaw)
	if err != nil {
		return false
	}

	digest := sha256.Sum256(payload)

	return ecdsa.Verify(public,
		digest[:],
		new(big.Int).SetBytes(signature[:SignatureLength/2]),
		new(big.Int).SetBytes(signature[SignatureLength/2:]),
	)
}

// signingKey reads the raw uncompressed point the browser exports. The parse rejects
// anything off the curve and the point at infinity, which is what closes the invalid-curve
// attack on the verifying side.
func signingKey(raw []byte) (*ecdsa.PublicKey, error) {
	public, err := ecdsa.ParseUncompressedPublicKey(elliptic.P256(), raw)
	if err != nil {
		return nil, fmt.Errorf("import signing key: %w", err)
	}

	return public, nil
}

func parseEC(pkcs8 []byte) (*ecdsa.PrivateKey, error) {
	parsed, err := x509.ParsePKCS8PrivateKey(pkcs8)
	if err != nil {
		return nil, fmt.Errorf("parse pkcs8: %w", err)
	}

	key, ok := parsed.(*ecdsa.PrivateKey)
	if !ok {
		return nil, fmt.Errorf("expected an EC key, got %T", parsed)
	}

	return key, nil
}

func readLengthPrefixed(bundle []byte, offset int) ([]byte, int, error) {
	if offset+2 > len(bundle) {
		return nil, 0, fmt.Errorf("identity bundle is truncated")
	}

	length := int(binary.BigEndian.Uint16(bundle[offset:]))
	start := offset + 2

	if start+length > len(bundle) {
		return nil, 0, fmt.Errorf("identity bundle is truncated")
	}

	return bundle[start : start+length], start + length, nil
}

// Identity is the pair of P-256 keypairs an account carries. One agrees on keys, one signs;
// reusing a single EC key for both is the kind of shortcut that turns into a break later.
type Identity struct {
	Seal *ecdh.PrivateKey
	Sign *ecdsa.PrivateKey

	// PublicBlob is what goes into users.public_key, and what a fingerprint is taken over.
	PublicBlob []byte
}

// GenerateIdentity produces an account identity in the exact shape the browser produces,
// so that a connector is indistinguishable from a person to everything that reads a grant.
func GenerateIdentity() (*Identity, error) {
	seal, err := ecdh.P256().GenerateKey(rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("generate agreement key: %w", err)
	}

	sign, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("generate signing key: %w", err)
	}

	blob := make([]byte, 0, 1+PublicKeyLength*2)
	blob = append(blob, IdentityFormat)
	blob = append(blob, seal.PublicKey().Bytes()...)

	signPublic, err := sign.PublicKey.ECDH()
	if err != nil {
		return nil, fmt.Errorf("encode signing key: %w", err)
	}

	blob = append(blob, signPublic.Bytes()...)

	return &Identity{Seal: seal, Sign: sign, PublicBlob: blob}, nil
}

// MarshalPrivateBundle packs both private keys the way unwrapIdentity expects to find them:
// the format byte, then each PKCS#8 key behind a big-endian uint16 length, agreement first.
func (id *Identity) MarshalPrivateBundle() ([]byte, error) {
	seal, err := x509.MarshalPKCS8PrivateKey(id.Seal)
	if err != nil {
		return nil, fmt.Errorf("marshal agreement key: %w", err)
	}

	sign, err := x509.MarshalPKCS8PrivateKey(id.Sign)
	if err != nil {
		return nil, fmt.Errorf("marshal signing key: %w", err)
	}

	bundle := make([]byte, 0, 1+2+len(seal)+2+len(sign))
	bundle = append(bundle, IdentityFormat)
	bundle = appendLengthPrefixed(bundle, seal)
	bundle = appendLengthPrefixed(bundle, sign)

	return bundle, nil
}

// Fingerprint is the short, readable digest of a public blob. The server hands out public
// keys, so it can hand out its own; comparing fingerprints out of band is what closes that
// gap, and it is the only check a person can perform on the connector's identity.
func Fingerprint(publicBlob []byte) string {
	digest := sha256.Sum256(publicBlob)

	return group(base32(digest[:], fingerprintChars), fingerprintGroup, ' ')
}

func appendLengthPrefixed(dst, value []byte) []byte {
	return append(binary.BigEndian.AppendUint16(dst, uint16(len(value))), value...)
}

// The additional data of the two blobs that wrap an account's keys. A connector account has
// the same shape as a person's: a master key wrapped by something derived from a secret,
// and the identity bundle wrapped by that master key. For a person the outer secret is a
// passphrase run through Argon2id; for a connector it is the server's configured secret.
const (
	masterKeyAAD = "shelf/master-key/v1"
	identityAAD  = "shelf/identity/v1"
)

func WrapMasterKey(masterKey, wrappingKey []byte) (Sealed, error) {
	return encrypt(wrappingKey, masterKey, []byte(masterKeyAAD))
}

func UnwrapMasterKey(sealed Sealed, wrappingKey []byte) ([]byte, error) {
	return decrypt(wrappingKey, sealed, []byte(masterKeyAAD))
}

func WrapIdentity(bundle, masterKey []byte) (Sealed, error) {
	return encrypt(masterKey, bundle, []byte(identityAAD))
}

// UnwrapIdentity reassembles an identity from what the database holds: the public blob as
// stored, and the bundle sealed under the master key.
func UnwrapIdentity(publicBlob []byte, sealed Sealed, masterKey []byte) (*Identity, error) {
	if _, _, err := SplitPublicBlob(publicBlob); err != nil {
		return nil, err
	}

	bundle, err := decrypt(masterKey, sealed, []byte(identityAAD))
	if err != nil {
		return nil, fmt.Errorf("unwrap identity: %w", err)
	}

	seal, sign, err := ParsePrivateBundle(bundle)
	if err != nil {
		return nil, err
	}

	return &Identity{Seal: seal, Sign: sign, PublicBlob: publicBlob}, nil
}
